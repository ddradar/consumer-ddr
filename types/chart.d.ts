export type PlayStyle =
  | 'SINGLE'
  | 'DOUBLE'
  | 'COUPLE'
  | '6-PANELS'
  | '3-PANELS'
  | 'STEP BATTLE'

export type Chart = {
  songId: string
  playStyle: PlayStyle
  difficulty: number
  level: number | '10+'
}
