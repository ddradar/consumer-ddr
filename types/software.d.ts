type Region = 'JP' | 'US' | 'EU' | 'None'

type Platform =
  | 'Play Station'
  | 'Play Station 2'
  | 'Dreamcast'
  | 'GAME BOY COLOR'
  | 'Wii'

export type Software = {
  id: string
  name: string
  region: Region
  platform: Platform
  launched: Date | null
}
