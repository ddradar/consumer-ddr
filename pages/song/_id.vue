<template>
  <section v-if="info" class="section">
    <h1 class="title">{{ info.name }} / {{ info.artist }}</h1>
    <h2 class="subtitle">BPM {{ info.bpm }}</h2>
    <div class="buttons">
      <b-button
        v-for="id in info.seriesList"
        :key="id"
        tag="nuxt-link"
        size="is-small"
        :class="getClass(id)"
        :to="`/series/${id}/`"
      >
        {{ id }}
      </b-button>
    </div>
    <b-table :data="info.charts" striped mobile-cards>
      <template slot-scope="props">
        <b-table-column field="playStyle" label="PlayStyle">
          {{ props.row.playStyle }}
        </b-table-column>
        <b-table-column field="difficulty" label="Difficulty">
          {{ props.row.difficulty }}
        </b-table-column>
        <b-table-column field="level" label="Level">
          {{ Object.values(props.row.levels).join(', ') }}
        </b-table-column>
      </template>
    </b-table>
  </section>
</template>

<script lang="ts">
import { Context } from '@nuxt/types'
import { Component, Vue } from 'nuxt-property-decorator'
import { MetaInfo } from 'vue-meta'

import { Chart, Song } from '~/types/song'

type ComplexChart = Omit<Chart, 'level'> & {
  levels: { [key in string]: number | '10+' | '?' }
}

type SongInfo = Omit<Song, 'series' | 'charts'> & {
  charts: ComplexChart[]
  seriesList: string[]
}

@Component
export default class SongDetailPage extends Vue {
  info?: SongInfo

  async asyncData({ params, $content }: Pick<Context, 'params' | '$content'>) {
    const songs: Song[] = await $content({ deep: true })
      .where({ slug: params.id })
      .fetch()
    const info = songs.reduce((prev, current) => {
      if (!prev) {
        return {
          name: current.name,
          artist: current.artist,
          bpm: current.bpm,
          charts: current.charts.map<ComplexChart>((c) => {
            const levels: { [key in string]: number | '10+' | '?' } = {}
            levels[current.series] = c.level
            return {
              playStyle: c.playStyle,
              difficulty: c.difficulty,
              levels
            }
          })
        } as SongInfo
      }
      prev.seriesList.push(current.series)
      for (const chart of current.charts) {
        const existsChart = prev.charts.find(
          (c) =>
            c.playStyle === chart.playStyle && c.difficulty === chart.difficulty
        )
        if (existsChart) {
          existsChart.levels[current.series] = chart.level
        } else {
          const levels: { [key in string]: number | '10+' | '?' } = {}
          levels[current.series] = chart.level
          prev.charts.push({
            playStyle: chart.playStyle,
            difficulty: chart.difficulty,
            levels
          })
        }
      }
      return prev
    }, undefined as SongInfo | undefined)
    return { info }
  }

  head(): MetaInfo {
    return {
      title: this.info?.name
    }
  }

  getClass(id: string) {
    const index = this.info!.seriesList.findIndex((s) => s === id)
    const classList = [
      'is-info',
      'is-success',
      'is-danger',
      'is-warning',
      'is-dark'
    ]
    return classList[index % classList.length]
  }
}
</script>
