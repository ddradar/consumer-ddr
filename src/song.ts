const basicNames = ['BASIC', 'NORMAL', 'LIGHT', '楽']
const difficultNames = ['ANOTHER', 'DIFFICULT', 'TRICK', '踊', 'STANDARD']
const expertNames = ['MANIAC', 'SSR', 'EXPERT', '激', 'HEAVY']
const beginnerNames = ['BEGINNER']

export function normalizeDifficulty(difficultyName: string) {
  const name = difficultyName.toUpperCase()
  if (basicNames.includes(name)) return 'BASIC'
  if (difficultNames.includes(name)) return 'DIFFICULT'
  if (expertNames.includes(name)) return 'EXPERT'
  if (beginnerNames.includes(name)) return 'BEGINNER'
  return 'CHALLENGE'
}
