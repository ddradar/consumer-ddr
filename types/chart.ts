type PlayStyle = 'SINGLE' | 'DOUBLE' | 'COUPLE' | '6-PANELS' | '3-PANELS'

export type Chart = {
  songId: string
  playStyle: PlayStyle
  difficulty: number
  level: number | '10+'
}
