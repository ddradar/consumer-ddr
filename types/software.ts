type Region = 'JP' | 'US' | 'EU'

type Platform = 'Play Station' | 'Play Station 2' | 'GAME BOY' | 'Wii'

export type Software = {
  id: string
  name: string
  region: Region
  platform: Platform
  launched: Date | null
}
