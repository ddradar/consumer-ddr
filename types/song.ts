export type Song = {
  slug?: string
  series: string
  name: string
  artist: string
  bpm: number | string
  charts: Chart[]
}

export type Chart = {
  playStyle: PlayStyle
  difficulty: number
  level: number | '10+' | '?'
}

export type PlayStyle =
  | 'SINGLE'
  | 'DOUBLE'
  | 'COUPLE'
  | 'UNISON'
  | '6-PANELS'
  | '3-PANELS'
  | 'STEP BATTLE'

export function normalizeDifficulty(difficultyName: string) {
  const name = difficultyName.toUpperCase()
  if (['BASIC', 'NORMAL', 'LIGHT'].includes(name)) return 'BASIC'
  if (['ANOTHER', 'DIFFICULT', 'TRICK'].includes(name)) return 'DIFFICULT'
  if (['MANIAC', 'SSR', 'EXPERT'].includes(name)) return 'EXPERT'
  if (['BEGINNER'].includes(name)) return 'BEGINNER'
  return 'CHALLENGE'
}
