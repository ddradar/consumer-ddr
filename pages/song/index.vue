<template>
  <section class="section">
    <h1 class="title">Song List</h1>

    <o-table :data="songList" per-page="20" striped narrowed paginated>
      <o-table-column v-slot="props" field="name" label="Name" searchable>
        <nuxt-link class="is-size-6-mobile" :to="`/song/${props.row.slug}/`">
          {{ props.row.name }}
        </nuxt-link>
      </o-table-column>
      <o-table-column v-slot="props" field="artist" label="Artist" searchable>
        <span class="is-size-6-mobile">{{ props.row.artist }}</span>
      </o-table-column>
      <o-table-column v-slot="props" field="bpm" label="BPM">
        {{ props.row.bpm }}
      </o-table-column>
      <o-table-column v-slot="props" label="Series">
        <div class="buttons">
          <o-button
            v-for="id in props.row.seriesList"
            :key="id"
            tag="nuxt-link"
            size="small"
            :variant="getClass(id)"
            :to="`/series/${id}/`"
          >
            {{ id }}
          </o-button>
        </div>
      </o-table-column>

      <template #empty>
        <section class="section">
          <div class="content has-text-grey has-text-centered">
            <p>Nothing here.</p>
          </div>
        </section>
      </template>
    </o-table>
  </section>
</template>

<script lang="ts">
import { Context } from '@nuxt/types'
import { Component, Vue } from 'nuxt-property-decorator'
import { MetaInfo } from 'vue-meta'

import { Song } from '~/types/song'

type SongListData = Omit<Song, 'series' | 'charts'> & {
  seriesList: string[]
}

@Component
export default class SeriesDetailPage extends Vue {
  songList: SongListData[] = []
  seriesList: string[] = []

  readonly head: MetaInfo = {
    title: 'Song List'
  }

  async asyncData({ $content }: Pick<Context, '$content'>) {
    const songs = await $content({ deep: true })
      .where({ extension: { $eq: '.json' } })
      .without('charts')
      .fetch<Omit<Song, 'charts'>>()
    const songList = [songs].flat().reduce((prev, current) => {
      if (!prev) {
        return [
          {
            slug: current.slug,
            name: current.name,
            artist: current.artist,
            bpm: current.bpm,
            seriesList: [current.series]
          }
        ]
      }
      if (!prev.some((s) => s.slug === current.slug)) {
        prev.push({
          slug: current.slug,
          name: current.name,
          artist: current.artist,
          bpm: current.bpm,
          seriesList: [current.series]
        })
      }
      return prev
    }, undefined as SongListData[] | undefined)

    const seriesList = (await $content({ deep: true })
      .where({ extension: { $eq: '.md' } })
      .sortBy('launched')
      .only('slug')
      .fetch<string>()) as string[]

    return { songList, seriesList }
  }

  getClass(id: string) {
    const index = this.seriesList!.findIndex((s) => s === id)
    const classList = ['info', 'success', 'danger', 'warning', 'dark']
    return classList[index % classList.length]
  }
}
</script>
