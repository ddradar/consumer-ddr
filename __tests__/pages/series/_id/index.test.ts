import {
  createLocalVue,
  mount,
  RouterLinkStub,
  shallowMount,
  Wrapper
} from '@vue/test-utils'
import Buefy from 'buefy'

import SeriesDetail from '~/pages/series/_id/index.vue'
import { Software } from '~/types/software'
import { Chart, PlayStyle, Song } from '~/types/song'

type ChartInfo = Pick<Chart, 'level' | 'difficulty'> & {
  color: string
  difficultyName: string
}

type SongListData = Omit<Song, 'series' | 'charts'> & {
  charts: [PlayStyle, ChartInfo[]][]
}

const localVue = createLocalVue()
localVue.use(Buefy)

describe('pages/series/_id/index.vue', () => {
  let wrapper: Wrapper<Vue>
  const info: Software = {
    slug: '1st-jp',
    name: 'Dance Dance Revolution',
    platform: 'Play Station',
    region: 'JP',
    launched: '1999-04-10',
    difficultyNames: {
      '1': 'BASIC',
      '2': 'ANOTHER',
      '3': 'MANIAC',
      '5': 'NORMAL'
    }
  }
  const songs: SongListData[] = [
    {
      slug: 'make-it-better',
      name: 'MAKE IT BETTER',
      artist: 'mitsu-O!',
      bpm: 118,
      charts: [
        [
          'SINGLE',
          [
            {
              difficulty: 1,
              level: 5,
              color: 'is-basic',
              difficultyName: 'BASIC'
            },
            {
              difficulty: 2,
              level: 6,
              color: 'is-difficult',
              difficultyName: 'ANOTHER'
            },
            {
              difficulty: 3,
              level: 8,
              color: 'is-expert',
              difficultyName: 'MANIAC'
            }
          ]
        ],
        [
          'DOUBLE',
          [
            {
              difficulty: 1,
              level: 5,
              color: 'is-basic',
              difficultyName: 'BASIC'
            },
            {
              difficulty: 2,
              level: 6,
              color: 'is-difficult',
              difficultyName: 'ANOTHER'
            }
          ]
        ]
      ]
    },
    {
      slug: 'paranoia',
      name: 'PARANOiA',
      artist: '180',
      bpm: 180,
      charts: [
        [
          'SINGLE',
          [
            {
              difficulty: 1,
              level: 6,
              color: 'is-basic',
              difficultyName: 'BASIC'
            },
            {
              difficulty: 2,
              level: 7,
              color: 'is-difficult',
              difficultyName: 'ANOTHER'
            },
            {
              difficulty: 3,
              level: 8,
              color: 'is-expert',
              difficultyName: 'MANIAC'
            }
          ]
        ],
        [
          'DOUBLE',
          [
            {
              difficulty: 1,
              level: 7,
              color: 'is-basic',
              difficultyName: 'BASIC'
            },
            {
              difficulty: 2,
              level: 8,
              color: 'is-difficult',
              difficultyName: 'ANOTHER'
            }
          ]
        ]
      ]
    },
    {
      slug: 'trip-machine',
      name: 'TRIP MACHINE',
      artist: 'DE-SIRE',
      bpm: 160,
      charts: [
        [
          'SINGLE',
          [
            {
              difficulty: 1,
              level: 6,
              color: 'is-basic',
              difficultyName: 'BASIC'
            },
            {
              difficulty: 2,
              level: 7,
              color: 'is-difficult',
              difficultyName: 'ANOTHER'
            },
            {
              difficulty: 3,
              level: 8,
              color: 'is-expert',
              difficultyName: 'MANIAC'
            }
          ]
        ],
        [
          'DOUBLE',
          [
            {
              difficulty: 1,
              level: 7,
              color: 'is-basic',
              difficultyName: 'BASIC'
            },
            {
              difficulty: 2,
              level: 8,
              color: 'is-difficult',
              difficultyName: 'ANOTHER'
            }
          ]
        ]
      ]
    }
  ]

  beforeEach(() => {
    wrapper = shallowMount(SeriesDetail, {
      localVue,
      data: () => ({ info, songs }),
      stubs: {
        NuxtLink: RouterLinkStub
      }
    })
  })

  test('renders correctly', () => {
    // Arrange - Act
    const wrapper = mount(SeriesDetail, {
      localVue,
      data: () => ({ info, songs }),
      stubs: {
        NuxtLink: RouterLinkStub
      }
    })

    // Assert
    expect(wrapper.element).toMatchSnapshot()
  })
  test('renders blank', () => {
    const wrapper = shallowMount(SeriesDetail, {
      localVue,
      data: () => ({ info: undefined }),
      stubs: {
        NuxtLink: RouterLinkStub
      }
    })
    expect(wrapper.html()).toBe('')
  })

  describe('head()', () => {
    test('returns info.name', () => {
      // Arrange
      const head = wrapper.vm.$options.head as any

      // Act & Assert
      expect(head.call(wrapper.vm)).toStrictEqual({ title: info.name })
    })
  })
})
