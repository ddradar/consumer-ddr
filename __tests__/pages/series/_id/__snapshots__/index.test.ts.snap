// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/series/_id/index.vue renders correctly 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    Dance Dance Revolution
  </h1>
   
  <h2
    class="subtitle"
  >
    Play Station(JP)
  </h2>
   
  <div
    class="b-table"
  >
    <!---->
     
    <!---->
     
    <!---->
     
    <!---->
    <!---->
    <!---->
    <!---->
    <div
      class="table-wrapper has-mobile-cards"
    >
      <table
        class="table is-striped is-narrow"
      >
        <thead>
          <tr>
            <!---->
            <!---->
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   Name 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   Artist 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   BPM 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   SINGLE 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   DOUBLE 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <!---->
          </tr>
          <!---->
          <!---->
        </thead>
        <tbody>
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="Name"
            >
              <a>
                
        MAKE IT BETTER
      
              </a>
            </td>
            <td
              class=""
              data-label="Artist"
            >
              <span>
                mitsu-O!
              </span>
            </td>
            <td
              class=""
              data-label="BPM"
            >
              
      118
    
            </td>
            <td
              class=""
              data-label="SINGLE"
            >
              <div
                class="tags has-addons"
              >
                <span
                  class="tag is-basic"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          BASIC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            5
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-difficult"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          ANOTHER
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            6
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-expert"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          MANIAC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            8
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
              </div>
            </td>
            <td
              class=""
              data-label="DOUBLE"
            >
              <div
                class="tags has-addons"
              >
                <span
                  class="tag is-basic"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          BASIC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            5
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-difficult"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          ANOTHER
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            6
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
              </div>
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="Name"
            >
              <a>
                
        PARANOiA
      
              </a>
            </td>
            <td
              class=""
              data-label="Artist"
            >
              <span>
                180
              </span>
            </td>
            <td
              class=""
              data-label="BPM"
            >
              
      180
    
            </td>
            <td
              class=""
              data-label="SINGLE"
            >
              <div
                class="tags has-addons"
              >
                <span
                  class="tag is-basic"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          BASIC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            6
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-difficult"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          ANOTHER
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            7
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-expert"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          MANIAC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            8
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
              </div>
            </td>
            <td
              class=""
              data-label="DOUBLE"
            >
              <div
                class="tags has-addons"
              >
                <span
                  class="tag is-basic"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          BASIC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            7
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-difficult"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          ANOTHER
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            8
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
              </div>
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="Name"
            >
              <a>
                
        TRIP MACHINE
      
              </a>
            </td>
            <td
              class=""
              data-label="Artist"
            >
              <span>
                DE-SIRE
              </span>
            </td>
            <td
              class=""
              data-label="BPM"
            >
              
      160
    
            </td>
            <td
              class=""
              data-label="SINGLE"
            >
              <div
                class="tags has-addons"
              >
                <span
                  class="tag is-basic"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          BASIC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            6
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-difficult"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          ANOTHER
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            7
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-expert"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          MANIAC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            8
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
              </div>
            </td>
            <td
              class=""
              data-label="DOUBLE"
            >
              <div
                class="tags has-addons"
              >
                <span
                  class="tag is-basic"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          BASIC
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            7
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
                <span
                  class="tag is-difficult"
                >
                  <span
                    class=""
                  >
                    <span
                      class="b-tooltip is-primary is-top is-medium"
                    >
                      <transition-stub
                        name="fade"
                      >
                        <div
                          class="tooltip-content"
                          style="display: none;"
                        >
                          ANOTHER
                        </div>
                      </transition-stub>
                      <div
                        class="tooltip-trigger"
                      >
                        
            8
          
                      </div>
                    </span>
                  </span>
                  <!---->
                </span>
              </div>
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <!---->
        </tbody>
        <!---->
      </table>
      <!---->
    </div>
    <!---->
  </div>
</section>
`;
